src/clr-angular/utils/focus/key-focus/key-focus.ts

```diff
@@ -0,0 +1,166 @@
+  get current() {
+    return this._current;
+  }
+
+  ngOnInit() {
+    if (!this.direction) {
+      this.direction = 'vertical';
+    }
+    if (this.focusOnLoad && isPlatformBrowser(this.platformId)) {
+      this.getCurrent().focus();
+      this._focusChange.next();
+    }
+  }
+
+  ngAfterContentChecked() {
+    this.getChildren().forEach(button => {
+      this.setTabIndex(button, -1);
+    });
+    this.setTabIndex(this.getCurrent(), 0);
+  }
+
+  @HostListener('keydown', ['$event'])
+  handleKeyboardEvent(event: any) {
+    if (this.prevClicked(event) && this._current - 1 >= 0) {
+      this.keyAction(() => {
+        this._current--;
+      });
+    } else if (this.nextClicked(event) && this._current + 1 < this.getChildren().length) {
+      this.keyAction(() => {
+        this._current++;
+      });
+    } else if (event.key === 'Home') {
+      this.keyAction(() => {
+        this._current = 0;
+      });
+    } else if (event.key === 'End') {
+      this.keyAction(() => {
+        this._current = this.getChildren().length - 1;
+      });
+    }
+  }
+
+  @HostListener('click', ['$event'])
+  handleClickEvent(event: any) {
+    const position = this.getPositionOf(event.srcElement as HTMLElement);
+    if (position > -1) {
+      this._current = position;
+    }
+  }
+
+  getPositionOf(element: HTMLElement) {
+    const children = Array.from(this.getChildren());
+    return children.indexOf(element);
+  }
+
+  moveTo(position: number) {
+    if (
+      isPlatformBrowser(this.platformId) &&
+      position >= 0 &&
+      position < this.getChildren().length &&
+      position !== this._current
+    ) {
+      this._current = position;
+      this.getCurrent().focus();
+      this._focusChange.next();
+    }
+  }
+
+  private keyAction(action: Function) {
+    if (!isPlatformBrowser(this.platformId)) {
+      return;
+    }
+    this.setTabIndex(this.getCurrent(), -1);
+    action.call(this);
+    this.setTabIndex(this.getCurrent(), 0);
+    this.getCurrent().focus();
+    this._focusChange.next();
+    // Prevent window scroll on arrow key pressed
+    event.preventDefault();
+    event.stopImmediatePropagation();
+  }
+
+  private prevClicked(event: any) {
+    switch (this.direction) {
+      case 'vertical':
+        return event.key === 'ArrowUp';
+      case 'horizontal':
+        return event.key === 'ArrowLeft';
+      case 'both':
+        return event.key === 'ArrowUp' || event.key === 'ArrowLeft';
+      default:
+        return false;
+    }
+  }
+
+  private nextClicked(event: any) {
+    switch (this.direction) {
+      case 'vertical':
+        return event.key === 'ArrowDown';
+      case 'horizontal':
+        return event.key === 'ArrowRight';
+      case 'both':
+        return event.key === 'ArrowDown' || event.key === 'ArrowRight';
+      default:
+        return false;
+    }
+  }
+
+  private getChildren(): HTMLElement[] {
+    return this.elementRef.nativeElement.querySelectorAll('button.btn') || [];
```

Comment:
Is it possible to use @ ContentChildren here? Asking because what if someone uses an `*ngIf` on one of the buttons. It would also remove the need for running code in every AfterContentChecked hook.

Reply:
With ContentChildren we would require the user to apply some directive to each of the child buttons with the sole purpose to help us achieve cleaner code :) 
I would like to avoid this if possible.
I also had several other points that helped me choose querySelectorAll over ContentChildren:
- we're able to apply this directive on top of existing code without imposing new requirements on the children
- this is a rendering type of DOM manipulation (vs. structural DOM manipulations). In my understanding structural DOM manipulations are much better candidates for ContentChildren and in-component handling as there you work directly with the ViewContanerRefs and TemplateRefs. 
While rendering type DOM manipulations require access to a nativeElement anyway and is pretty much happening on the DOM level itself and these type of manipulations don't even have a representation in the View layer. It doesn't make sense to go through the View layer just to get access to a nativeElement and do the actual work there.
- In addition to the above, the focus() handling is even more natively DOM-coupled, as you even don't have a renderer2 wrapper for it.

So, it's between you and the DOM alone. You don't need to get the View invited to this party, as he'll feel kinda isolated :D 

About the AfterContentChecked - IMO, if it happens way too often, then we need to investigate why it happens. Though the components that I am interested in have no issue with this, I also have a specific component in mind where I can investigate a bit deeper. I am open to discussing cases where constant content-checking is permissible and unavoidable.

I am open to deeper discussion and explanation of the structural vs rendering stuff too.

Reply:
That makes sense. Thank you for explaining the rational behind the choice.